### ISSUE TITLE
try_top_auth to send realm info

### ISSUE THREAD

**Issue by tcavalca** [adrienverge/openfortivpn#411]
First, thanks for the excellent work

I have implemented 2FA using Radius+PrivacyIdea and is working quite well.
For openfortivpn, I have found that "try_otp_auth()" does not send 'realm' information, so connection is dropped even if authentication was successful.

I've made a small patch to fix this - I have not coded for ages so I will appreciate a lot if you guys could come with a more elegant solution.

Let me know if you more information on this

Again, thanks for all the work
Tenorio

PS: adding the patch here for the curious, sorry if I was not supposed to do that
--------- patch begin -----------------
$ cat  http.patch 
--- src/http.c	2018-12-07 13:18:47.516683443 -0400
+++ src/http.c-realm	2018-12-07 16:27:44.840616228 -0400
@@ -500,6 +500,16 @@
 				return -1;
 			url_encode(d, cfg->otp);
 			d += strlen(d);
+			/*  realm workaround */
+			if (cfg->realm[0] != '\0') {
+				l = strlen(cfg->realm);
+				if (!SPACE_AVAILABLE(3 * l + 1))
+					return -1;
+				strcat(d, "&realm=");
+				d += strlen(d);
+				url_encode(d, cfg->realm);
+				d += strlen(d);
+			}
 		} else if (strncmp(t, "submit", 6) == 0) {
 			/* avoid adding another '&' */
 			n = v = e = NULL;
--------- patch end ---------------

**Comment by tcavalca** [comment]
Ops, forgot of markup lang ...

```
$ cat  http.patch 
--- src/http.c	2018-12-07 13:18:47.516683443 -0400
+++ src/http.c-realm	2018-12-07 16:27:44.840616228 -0400
@@ -500,6 +500,16 @@
 				return -1;
 			url_encode(d, cfg->otp);
 			d += strlen(d);
+			/*  realm workaround */
+			if (cfg->realm[0] != '\0') {
+				l = strlen(cfg->realm);
+				if (!SPACE_AVAILABLE(3 * l + 1))
+					return -1;
+				strcat(d, "&realm=");
+				d += strlen(d);
+				url_encode(d, cfg->realm);
+				d += strlen(d);
+			}
 		} else if (strncmp(t, "submit", 6) == 0) {
 			/* avoid adding another '&' */
 			n = v = e = NULL;
```

**Comment by mrbaseman** [comment]
posting the patch here is totally ok. 
I have just edited your first post and removed the code which was not properly formatted. Actually, instead of posting it again in a separate comment you could have edited the first one as well (as long as the changes are small corrections like formatting, spelling, grammar,...). 
About the patch itself: Thanks for reporting this, and we'll have a look. I think length check should not only consider the possible maximum length of the url encoded version of the realm, but additionally for the length of `"&realm="`, so it should read `SPACE_AVAILABLE(3 * l + 8)` i.m.o.

**Comment by tcavalca** [comment]
yes, you completely right - it should be `SPACE_AVAILABLE(3 * 1 + 8)`
